### PR TITLE
Fix coupon offers not being set with correct partner.

### PR DESCRIPTION
**Pull request by waheedahmed**
LEARNER-6393

### CODE REVIEW

File: ecommerce/extensions/voucher/tests/test_utils.py [hunk 1]
```diff
@@ -200,6 +201,7 @@ def test_create_vouchers(self):
         self.assertEqual(voucher_offer.benefit.range.catalog, self.catalog)
         self.assertEqual(voucher_offer.email_domains, email_domains)
         self.assertEqual(voucher_offer.priority, OFFER_PRIORITY_VOUCHER)
+        self.assertEqual(voucher_offer.partner, self.partner)
```
Comment (robrap) [inline ecommerce/extensions/voucher/tests/test_utils.py:204]: Can you confirm that this test covers the actual change, and not just the factory change? If not, is there a way to get a real test?

**Review by robrap — COMMENTED** [review]
I guess we missed this. Looks good, but one question. Also, can you confirm that you were able to manually test this?

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 1 above)
Comment (waheedahmed) [reply ecommerce/extensions/voucher/tests/test_utils.py:204]: @robrap Let me try with a real test. Yes, I tested it manually by creating a coupon on my local machine and partner was successfully added to offers.

File: ecommerce/extensions/test/factories.py [hunk 2]
```diff
@@ -122,7 +122,7 @@ def prepare_voucher(code='COUPONTEST', _range=None, start_datetime=None, end_dat
             benefit=benefit,
             condition=condition,
             email_domains=email_domains,
-            site=site,
```
Comment (staubina) [inline ecommerce/extensions/test/factories.py:125]: Can we set the site too for safety? Or would that cause other issues?

Comment (waheedahmed) [reply ecommerce/extensions/test/factories.py:125]: @staubina We are going to remove site object completely and using partner instead.

Comment (robrap) [reply ecommerce/extensions/test/factories.py:125]: Agreed. Since we switched from site offers to partner offers, I think leaving the site could just lead to other future bugs and misunderstanding.

**Comment by robrap** [comment]
I can’t yet see the Travis failures. I generally approve, but want a green build first. We may leave an additional test to a follow-up PR.

**Comment by bill-filler** [comment]
@waheedahmed can you grep the code to make sure there are not any other places where we are creating the ConditionalOffer and not correctly setting the partner? I want to make sure we haven't missed any other cases.

File: ecommerce/extensions/test/factories.py (hunk 2 above)
Comment (staubina) [reply ecommerce/extensions/test/factories.py:125]: Ok, we should update the white label team as well.

@bill-filler FYI

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 1 above)
Comment (waheedahmed) [reply ecommerce/extensions/voucher/tests/test_utils.py:204]: @robrap `create_vouchers ` is not a factory function it is the actual function which is used to create vouchers and offers while creating coupons.

**Comment by waheedahmed** [comment]
@bill-filler Yes, I did grep the code and found this place only and a test case where conditional offer is being created.

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 1 above)
Comment (robrap) [reply ecommerce/extensions/voucher/tests/test_utils.py:204]: Great. I wasn’t sure from just the PR code. 

Still may want to consider a deeper integration test of creating an offer and ensuring it applies appropriately, akin to what you tested manually.

File: ecommerce/extensions/voucher/utils.py [hunk 3]
```diff
@@ -373,7 +373,7 @@ def _get_or_create_offer(
         benefit=offer_benefit,
         max_global_applications=max_uses,
         email_domains=email_domains,
-        site=site,
+        partner=site.siteconfiguration.partner if site else None,
```
Comment (robrap) [inline ecommerce/extensions/voucher/utils.py]: What would happen if Partner were None? Is that ever valid, or do we need to log an error?

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 1 above)
Comment (waheedahmed) [reply ecommerce/extensions/voucher/tests/test_utils.py:204]: @robrap Sure thing, I'll try to add a test case in follow-up PR.

File: ecommerce/extensions/test/factories.py (hunk 2 above)
Comment (waheedahmed) [reply ecommerce/extensions/test/factories.py:125]: @robrap So, should I set the site as well for now?

**Comment by bill-filler** [comment]
Also wondering if we need a migration similar to the previous PR which updates the partner ID field for all the conditional offers. The use case would be to fix all of the offers that have been created since the original PR landed all have null partner ID currently

File: ecommerce/extensions/voucher/utils.py [hunk 4]
```diff
@@ -374,6 +374,7 @@ def _get_or_create_offer(
         max_global_applications=max_uses,
         email_domains=email_domains,
         site=site,
```
Comment (waheedahmed) [inline ecommerce/extensions/voucher/utils.py:376]: @staubina @robrap Added the site back.

Comment (staubina) [reply ecommerce/extensions/voucher/utils.py:376]: I thought you an @robrap  wanted to remove it.

**Comment by staubina** [comment]
> Also wondering if we need a migration similar to the previous PR which updates the partner ID field for all the conditional offers. The use case would be to fix all of the offers that have been created since the original PR landed all have null partner ID currently

@waheedahmed Please consider this request as well.  We have a number of items without partner right now that we might want to fix.

**Comment by waheedahmed** [comment]
@staubina working on it.

**Comment by waheedahmed** [comment]
@bill-filler, @robrap, @staubina Added the migration to update partner for offers created in Sep, 2018. Kindly take a look.

**Review by bill-filler — APPROVED** [review]
looks good, thanks

File: ecommerce/extensions/voucher/utils.py (hunk 4 above)
Comment (robrap) [reply ecommerce/extensions/voucher/utils.py:376]: @staubina @waheedahmed: I’ll Let you both decide what is safest in the short term. Once the dust has settled, there should be a follow-up clean up if you’ve decided it is safer to leave it for now.

**Review by robrap — COMMENTED** [review]
Does the migration need to go separately, or do we just need to ensure no test coupons come in on the old cluster during deployment? At a minimum, we should check the db afterwards to ensure all is well.

**Comment by waheedahmed** [comment]
@robrap Actually that is a valid concern, let me make a separate PR for the migration which lands on production after this PR. Does this sounds good?

**Comment by waheedahmed** [comment]
@robrap Created the separate PR for the migration https://github.com/edx/ecommerce/pull/1912.

CC: @bill-filler @staubina
